### DISCUSSION TITLE
How to exclude columns when inserting from the client side

### DISCUSSION THREAD

**Discussion by mozeryansky** [supabase/supabase#6560] [Questions] [answered]
For tables which have primary keys and default columns like created/update/etc, how can I prevent client api calls from being able to provide values for these columns when inserting a row?

I rather not have to create rpc functions for all my inserts. Right now the security is done by only allowing a user to insert rows with their own user_id and the primary key has unique constraints, but that doesn't prevent a user from specifying an id themselves or manually setting the created date.

**Comment by GaryAustin1** [comment, marked as answer]
Postgres has column level security by user/role:  https://github.com/supabase/supabase/discussions/4744

You can also have a before insert/update trigger function and set all your defaults using new.col1 = now(), or even not allow updates by doing new.col2 = old.col2 using TG_OP to detect insert or delete.   General info on using TG_OP,new, old can be found here: https://www.postgresql.org/docs/current/plpgsql-trigger.html
If you go the trigger route then even super user won't be able to update, unless you have if statements on role.

**Reply by mozeryansky** [reply]
I've tested this out to verify and column level security works great. One thing I learned is that revoke only removes already granted privileges. "The REVOKE command revokes previously granted privileges from one or more roles." [postgresql.org docs](https://www.postgresql.org/docs/14/sql-revoke.html)

So the steps are:
1. REVOKE INSERT ON my_table FROM anon
2. GRANT INSERT(field_1, field_2) ON my_table TO anon
And you will need to consider this for the 'anon' and 'authenticated' users.

Now when I attempt an insert from the client and include the `id` column I get the error "permission denied for table test". When I insert without the `id` column then the insert works and I safely get auto incremented ids.

**Reply by mozeryansky** [reply]
You can view all full grants using (which excludes grants which specify columns):
\> SELECT * FROM information_schema.role_table_grants WHERE table_name = 'my_table' AND grantee = 'anon'

And to see all the individual column privileges use:
\> SELECT * FROM information_schema.column_privileges WHERE table_name = 'my_table' AND grantee = 'anon'

But it was much easier to use PgAdmin or pgsql to view the generated SQL for the table.

<img width="1309" alt="Screen Shot 2022-04-29 at 11 40 32 PM" src="https://user-images.githubusercontent.com/3287719/166091308-0305f04c-f337-4ac2-9b38-2a30caa98fbb.png">
